blueprint/core/blueprint_ReactApplicationRoot.h

```diff
@@ -392,6 +393,59 @@ namespace blueprint
          **/
         BundleEvalCallback afterBundleEval;
 
+        //==============================================================================
+        using EventHandler = std::function<void(const juce::var::NativeFunctionArgs& args)>;
+
+        struct EventSubscription
+        {
+            juce::String eventType;
+            size_t       subscriptionIndex{};
+            EventHandler handler;
+        };
+
+        /**
+         * Subscribe to a given event from js. This should be called in your beforeBundleEval
+         * callback as the set of listeners/subscriptions will be cleared on bundle reloads.
+         * Returns an EventSubscription instance which callers may store and later use to unsubscribe()
+         * a previously registered EventHandler function.
+         * Note, callers should take care to avoid callback loops between subscribe and dispatchEvent etc.
+         **/
+         EventSubscription subscribe(const juce::String &eventType, EventHandler &&handler)
+        {
+            JUCE_ASSERT_MESSAGE_THREAD
+            jassert(handler);
+
+            const size_t count = std::count_if(eventHandlerSubscriptions.begin(),
+                                               eventHandlerSubscriptions.end(),
+                                               [=] (const EventSubscription & sub)
+                                               {
+                                                   return sub.eventType == eventType;
+                                               });
+
+            EventSubscription sub = {eventType, count, std::move(handler)};
+            eventHandlerSubscriptions.push_back(sub);
```

Comment:
I'm wondering if a different data structure here would make the bookkeeping easier here. E.g. a `std::map<juce::Uuid, EventHandler>`?

```
EventToken subscribe(const juce::String& type, EventHandler && handler) {
    JUCE_ASSERT_MESSAGE_THREAD
    juce::Uuid token; // default construct
    eventSubscriptions[token] = std::move(handler):
    return token;
}
```

Reply:
Yeah I debated whether we should have a map here. Happy to switch things around that way.

Reply:
If we're storing a `juce::Uuid` token/key we probably still want a struct to store the `eventType` string and `EventHandler` together so we can call the correct handlers in our native `dispatchEventToNative` implementation?

Reply:
i.e. 

```
 using EventHandler = std::function<void(const juce::var::NativeFunctionArgs& args)>;
        using EventSubscriptionToken = juce::Uuid;

        struct EventSubscription
        {
            juce::String eventType;
            EventHandler handler;
        };

        /**
         * Subscribe to a given event from js. This should be called in your beforeBundleEval
         * callback as the set of listeners/subscriptions will be cleared on bundle reloads.
         * Returns an EventSubscriptionToken instance which callers may store and later use to unsubscribe()
         * a previously registered EventHandler.
         * Note, callers should take care to avoid callback loops between subscribe and dispatchEvent etc.
         **/
        EventSubscriptionToken subscribe(const juce::String &eventType, EventHandler &&handler)
        {
            JUCE_ASSERT_MESSAGE_THREAD
            jassert(handler);

            EventSubscriptionToken token;
            EventSubscription sub = {eventType, std::move(handler)};
            eventHandlerSubscriptions[token] = std::move(sub);

            return token;
        }
```

What dya think?

Reply:
```
 void registerNativeEventHandlers()
        {
            engine.registerNativeMethod("__BlueprintNative__", "dispatchEventToNative", [](void* stash, const juce::var::NativeFunctionArgs& args) {
                auto self = reinterpret_cast<ReactApplicationRoot*>(stash);
                jassert (self != nullptr);
                jassert (args.numArguments >= 1);

                const juce::String eventType = args.arguments[0].toString();

                std::for_each(self->eventHandlerSubscriptions.begin(),
                              self->eventHandlerSubscriptions.end(),
                              [&] (const EventSubscription &sub)
                              {
                                  if (sub.eventType == eventType)
                                  {
                                      sub.handler(juce::var::NativeFunctionArgs(
                                                      juce::var(),
                                                      args.arguments + 1,
                                                      args.numArguments - 1));
                                  }
                              });

                return juce::var::undefined();
            }, (void *) this);
        }
```

For reference

Reply:
Oh, geez, of course. Good catch, sorry about that!

Actually, somehow reading that just made me realize there's an even easier way:

```js
const EventBridge = new EventEmitter();

EventBridge.setMaxListeners(30);

// An internal hook for the native side, from which we propagate events through
// the EventEmitter interface.
__BlueprintNative__.dispatchEvent = function dispatchEvent(eventType, ...args) {
  EventBridge.emit(eventType, ...args);
}

__BlueprintNative__.subscribe = function subscribe(eventType, callback) {
    EventBridge.addListener(eventType, callback);
});

__BlueprintNative__.unsubscribe = function subscribe(eventType, callback) {
    EventBridge.removeListener(eventType, callback);
});

export default EventBridge;
```

Then `ReactApplicationRoot` only needs this:

```cpp
void subscribe (const juce::String& eventType, EventHandler && handler) {
    engine.invoke("__BlueprintNative__.subscribe", handler);
});
```

Because of the lambdaReleasePool stuff, this should all be able to live just fine in the javascript side of things with way less code. I'm hand-waving a little bit over the unsubscribe part– I think we'll need return tokens on that side. I can elaborate there if it's helpful, but what do you think?

Reply:
Ugh yeah of course. I didn't even think about the fact that we could stash the lambda like that. That would be neat. 

So in `EventBridge.unsubscribe` we could just return a counter/token and store a ref to the callback function in JS ?

Reply:
Exactly!

Reply:
Ha flipping eck. That is so much more elegant. I expect as I continue working with Blueprint I'll spot these patterns more often but this gets rid of all the `ToNative` `FromNative` as well. Jolly good catch sir.

Reply:
I'm trying to decide whether unsubscribe is actually essential ... it's the only bit of ugliness here. I suppose there could be valid reasons for dynamic subscribe/unsubscribe to the events, it's not something I need at the moment personally. What do you think? We probably should have it ...

Reply:
Listeners will be cleared on a bundle reload anyway as the release pool and ctx get cleared out

Reply:
It would be nice to be able to keep the expected signature:

```
__BlueprintNative__.unsubscribe = function unsubscribe(eventType, callback) {
    EventBridge.removeListener(eventType, callback);
});
```

I understand easily enough you could return an int counter or whatever from `subscribe` and change this signature to:

```
__BlueprintNative__.unsubscribe = function unsubscribe(token) {
    let listener = listeners[token];
    EventBridge.removeListener(listener.eventType, listener.callback);
});
```

Is there a better way to do this?

Reply:
Yea I think it's worth having an unsubscribe call. And I think we'll have to go the `token` approach because, at least now, passing the same lambda into `EcmascriptEngine::invoke` will just wrap it in a new function in the duktape context, so you won't be able to do `listener === cb` and get a `true` ever. I think what you wrote above there is plenty good.

```
EventBridge.__listeners = {};
EventBridge.__nextToken = 0;

__BlueprintNative__.subscribe = function(eventType, handler) {
    EventBridge.addListener(eventType, handler);

    const token = EventBridge.__nextToken++;
    EventBridge.__listeners[token] = handler;
    return token;
};
```
Then use exactly your `unsubscribe` up there.

Reply:
Yeah cool. I was musing over whether there was any smart way to pull the relevant lambda out of the duk stack etc but that sounds like either voodoo or an antipattern. Cool. Will get this tidied up when I'm back at my desk later. Cheers

Reply:
Awesome, thanks. Yea, I'm sure there's a way to pull it out of the duk stack but I feel like it would be way more complicated than the token pattern here

Reply:
@nick-thompson, OK hopefully the PR now fits our new expectations. Let me know if that's all good and I can squash it all into a single commit.